File: phases.py
"""各阶段处理逻辑

阶段流程：
  NIGHT → wolf_kill → seer_check → witch_action → NIGHT_END
  DAY   → day_announce → hunter_shoot(若触发) → speech → vote → vote_result
"""
from __future__ import annotations
import logging
import random
from typing import List, Optional, Dict, Any, TYPE_CHECKING

from models import (
    PlayerState, Role, NightAction, DeathRecord, DeathReason, Phase
)
from player import AIPlayer
from parser import parse_action, parse_speech, parse_witch_action
import prompts

if TYPE_CHECKING:
    from game import GameEngine

logger = logging.getLogger("ai-werewolf")


# ─── 夜晚阶段 ───

def phase_wolf_kill(engine: "GameEngine") -> None:
    """狼人商量杀人"""
    log = engine.log
    day = log.day
    night = NightAction(day=day)

    wolves = [p for p in engine.players if p.alive and p.role == Role.WEREWOLF]
    if not wolves:
        log.nights.append(night)
        return

    alive_ids = [p.state.player_id for p in engine.players if p.alive and p.role != Role.WEREWOLF]
    if not alive_ids:
        log.nights.append(night)
        return

    night_history = [
        {"day": n.day, "deaths": [{"name": engine.get_player_name(d.player_id)} for d in n.deaths]}
        for n in log.nights
    ]

    # 每个狼人独立投票
    votes: Dict[int, int] = {}
    for wolf in wolves:
        sys_p, user_p = prompts.wolf_kill_prompt(
            wolf.state, [p.state for p in engine.players], night_history,
            personality=wolf.state.personality,
        )
        raw = wolf.chat(sys_p, user_p, verbose=engine.verbose)
        target = parse_action(raw, "kill", alive_ids)

        # dry-run fallback: 随机选一个合法目标
        if target is None and engine.dry_run and alive_ids:
            target = random.choice(alive_ids)

        if target is not None:
            votes[wolf.id] = target

        if engine.verbose:
            logger.info(f"  🐺 {wolf.name} 选择杀: {target}")

    # 多数决（平局随机选）
    if votes:
        tally: Dict[int, int] = {}
        for t in votes.values():
            tally[t] = tally.get(t, 0) + 1
        max_count = max(tally.values())
        candidates = [t for t, c in tally.items() if c == max_count]
        night.wolf_target = random.choice(candidates)
    else:
        night.wolf_target = random.choice(alive_ids)

    if engine.verbose:
        name = engine.get_player_name(night.wolf_target)
        logger.info(f"  🌙 第{day}夜 狼人决定杀害: {name}({night.wolf_target})")

    engine.current_night = night


def phase_seer_check(engine: "GameEngine") -> None:
    """预言家查验"""
    night = engine.current_night
    if not night:
        return

    seer = _find_alive_player(engine, Role.SEER)
    if not seer:
        return

    alive_ids = [p.state.player_id for p in engine.players
                 if p.alive and p.state.player_id != seer.id]

    sys_p, user_p = prompts.seer_check_prompt(
        seer.state, [p.state for p in engine.players],
        engine.seer_results, [],
        personality=seer.state.personality,
    )
    raw = seer.chat(sys_p, user_p, verbose=engine.verbose)
    target = parse_action(raw, "check", alive_ids)

    # dry-run fallback
    if target is None and engine.dry_run and alive_ids:
        target = random.choice(alive_ids)

    if target is not None:
        night.seer_target = target
        target_player = engine.get_player_by_id(target)
        is_wolf = target_player.role == Role.WEREWOLF
        night.seer_result = "狼人" if is_wolf else "好人"

        engine.seer_results.append({
            "id": target,
            "name": target_player.name,
            "is_wolf": is_wolf,
        })

        if engine.verbose:
            logger.info(
                f"  🔮 预言家{seer.name}查验: {target_player.name}({target}) → {night.seer_result}"
            )


def phase_witch_action(engine: "GameEngine") -> None:
    """女巫行动（救/毒）"""
    night = engine.current_night
    if not night:
        return

    witch = _find_alive_player(engine, Role.WITCH)
    if not witch:
        return

    has_antidote = witch.state.witch_has_antidote
    has_poison = witch.state.witch_has_poison

    if not has_antidote and not has_poison:
        return

    killed_name = engine.get_player_name(night.wolf_target) if night.wolf_target else "未知"
    killed_id = night.wolf_target if night.wolf_target else -1

    sys_p, user_p = prompts.witch_action_prompt(
        witch.state, [p.state for p in engine.players],
        killed_name, killed_id, has_antidote, has_poison, [],
        personality=witch.state.personality,
    )
    raw = witch.chat(sys_p, user_p, verbose=engine.verbose)
    action = parse_witch_action(raw)

    # 救人
    if action["save"] and has_antidote and night.wolf_target:
        night.witch_saved = True
        witch.state.witch_has_antidote = False
        if engine.verbose:
            logger.info(f"  💊 女巫{witch.name}使用解药救了 {killed_name}")

    # 毒人
    alive_ids = [p.state.player_id for p in engine.players
                 if p.alive and p.state.player_id != witch.id]
    poison_target = action.get("poison")
    if poison_target and has_poison:
        poison_id = _to_int_safe(poison_target)
        if poison_id is not None and poison_id in alive_ids:
            night.witch_poisoned = poison_id
            witch.state.witch_has_poison = False
            if engine.verbose:
                poison_name = engine.get_player_name(poison_id)
                logger.info(f"  ☠️ 女巫{witch.name}使用毒药毒了 {poison_name}")


def phase_night_resolve(engine: "GameEngine") -> List[DeathRecord]:
    """结算夜晚死亡"""
    night = engine.current_night
    if not night:
        return []

    deaths = []

    # 狼人杀人（除非被女巫救）
    if night.wolf_target and not night.witch_saved:
        target = engine.get_player_by_id(night.wolf_target)
        if target and target.alive:
            target.alive = False
            death = DeathRecord(
                player_id=night.wolf_target,
                reason=DeathReason.WOLF_KILL,
                day=night.day,
                details=f"被狼人杀害",
            )
            deaths.append(death)

    # 女巫毒杀
    if night.witch_poisoned:
        target = engine.get_player_by_id(night.witch_poisoned)
        if target and target.alive:
            target.alive = False
            death = DeathRecord(
                player_id=night.witch_poisoned,
                reason=DeathReason.WITCH_POISON,
                day=night.day,
                details=f"被女巫毒杀",
            )
            deaths.append(death)

    night.deaths = deaths
    engine.log.nights.append(night)
    engine.log.deaths.extend(deaths)
    return deaths


# ─── 白天阶段 ───

def phase_day_announce(engine: "GameEngine", deaths: List[DeathRecord]) -> str:
    """宣布死讯"""
    if deaths:
        names = "、".join(
            f"{engine.get_player_name(d.player_id)}({d.details})" for d in deaths
        )
        msg = f"昨晚，{names}。"
    else:
        msg = "昨晚是平安夜，没有人死亡。"

    if engine.verbose:
        logger.info(f"  ☀️ {msg}")
    return msg


def phase_hunter_shoot(engine: "GameEngine", deaths: List[DeathRecord]) -> Optional[int]:
    """猎人被杀时开枪"""
    for death in deaths:
        player = engine.get_player_by_id(death.player_id)
        if player and player.role == Role.HUNTER and player.hunter_can_shoot:
            hunter_player = _find_player_by_id(engine, death.player_id)
            if not hunter_player:
                continue

            alive_ids = [p.state.player_id for p in engine.players if p.alive]
            reason_str = death.details

            sys_p, user_p = prompts.hunter_shoot_prompt(
                hunter_player.state, [p.state for p in engine.players],
                reason_str, [],
                personality=hunter_player.state.personality,
            )
            raw = hunter_player.chat(sys_p, user_p, verbose=engine.verbose)
            target = parse_action(raw, "shoot", alive_ids)

            if target is not None:
                target_player = engine.get_player_by_id(target)
                if target_player and target_player.alive:
                    target_player.alive = False
                    shoot_death = DeathRecord(
                        player_id=target,
                        reason=DeathReason.HUNTER_SHOOT,
                        day=engine.log.day,
                        details=f"被猎人{player.name}开枪射杀",
                    )
                    engine.log.deaths.append(shoot_death)
                    if engine.verbose:
                        logger.info(
                            f"  🔫 猎人{player.name}开枪射杀了 {target_player.name}({target})"
                        )
                    return target
    return None


def phase_speech(engine: "GameEngine", day: int, deaths_info: str) -> List[Dict[str, str]]:
    """白天发言阶段"""
    speeches = []
    alive_players = [p for p in engine.players if p.alive]

    for player in alive_players:
        # 构建私密信息（严格信息隔离）
        personal_info = ""
        if player.role == Role.SEER:
            personal_info = prompts.seer_personal_info(engine.seer_results)
        elif player.role == Role.WEREWOLF:
            wolves = [p.state for p in engine.players
                      if p.role == Role.WEREWOLF and p.id != player.id and p.alive]
            personal_info = prompts.wolf_personal_info(wolves)

        sys_p, user_p = prompts.speech_prompt(
            player.state, [p.state for p in engine.players],
            day, deaths_info, speeches, personal_info,
            personality=player.state.personality,
        )
        raw = player.chat(sys_p, user_p, verbose=engine.verbose)
        speech_text = parse_speech(raw)

        speeches.append({
            "player_id": player.id,
            "name": player.name,
            "content": speech_text,
        })

        if engine.verbose:
            logger.info(f"  🎤 [{player.id}] {player.name}: {speech_text[:80]}...")

    engine.log.speeches.append(speeches)
    return speeches


def phase_vote(engine: "GameEngine", day: int,
               speeches: List[Dict[str, str]]) -> "VoteResult":
    """投票阶段"""
    from models import VoteResult

    alive_players = [p for p in engine.players if p.alive]
    alive_ids = [p.state.player_id for p in alive_players]

    votes: Dict[int, int] = {}
    for player in alive_players:
        # 私密信息
        personal_info = ""
        if player.role == Role.SEER:
            personal_info = prompts.seer_personal_info(engine.seer_results)
        elif player.role == Role.WEREWOLF:
            wolves = [p.state for p in engine.players
                      if p.role == Role.WEREWOLF and p.id != player.id and p.alive]
            personal_info = prompts.wolf_personal_info(wolves)

        sys_p, user_p = prompts.vote_prompt(
            player.state, [p.state for p in engine.players],
            day, speeches, personal_info,
            personality=player.state.personality,
        )
        raw = player.chat(sys_p, user_p, verbose=engine.verbose)

        # 不能投自己
        valid_targets = [pid for pid in alive_ids if pid != player.id]
        target = parse_action(raw, "vote", valid_targets)

        # dry-run fallback
        if target is None and engine.dry_run and valid_targets:
            target = random.choice(valid_targets)

        if target is not None:
            votes[player.id] = target

        if engine.verbose:
            target_name = engine.get_player_name(target) if target else "无效"
            logger.info(f"  🗳️ [{player.id}] {player.name} → 投票给 {target_name}")

    # 计票
    tally: Dict[int, int] = {}
    for target in votes.values():
        tally[target] = tally.get(target, 0) + 1

    eliminated = None
    is_tie = False
    if tally:
        max_count = max(tally.values())
        candidates = [t for t, c in tally.items() if c == max_count]
        if len(candidates) == 1:
            eliminated = candidates[0]
        else:
            is_tie = True
            eliminated = random.choice(candidates)  # 平局随机

    # 执行放逐
    if eliminated is not None:
        target_player = engine.get_player_by_id(eliminated)
        if target_player:
            target_player.alive = False
            death = DeathRecord(
                player_id=eliminated,
                reason=DeathReason.VOTE_OUT,
                day=day,
                details=f"被投票放逐({tally.get(eliminated, 0)}票)",
            )
            engine.log.deaths.append(death)

            if engine.verbose:
                name = target_player.name
                logger.info(f"  ⚖️ {name}({eliminated}) 被投票放逐 ({tally.get(eliminated, 0)}票)")

    result = VoteResult(
        day=day, votes=votes, tally=tally,
        eliminated=eliminated, is_tie=is_tie,
    )
    engine.log.votes.append(result)
    return result


# ─── 辅助 ───

def _find_alive_player(engine: "GameEngine", role: Role) -> Optional[AIPlayer]:
    for p in engine.players:
        if p.alive and p.role == role:
            return p
    return None


def _find_player_by_id(engine: "GameEngine", pid: int) -> Optional[AIPlayer]:
    for p in engine.players:
        if p.id == pid:
            return p
    return None


def _to_int_safe(val) -> Optional[int]:
    if val is None:
        return None
    try:
        return int(val)
    except (ValueError, TypeError):
        return None

File: lib/engine/phases.ts
/**
 * 各阶段处理逻辑 —— 对应原 Python 版 phases.py
 *
 * 流程：
 *   夜晚: guard_protect → wolf_kill → seer_check → witch_action → night_resolve
 *   白天: day_announce → hunter_shoot(若触发) → speech → vote → hunter_shoot(若被投)
 *
 * 所有函数都是 async，phase 内多次 await LLM 调用。
 * 关键节点通过 engine.emitter 发射事件，供 SSE / DB / 统计消费。
 */
import type { GameEngine } from './game';
import type {
  DeathRecord,
  NightAction,
  PlayerState,
  VoteResult,
} from '../types';
import { DEATH_REASON } from '../types';
import type { AIPlayer } from '../llm/player';
import {
  parseAction,
  parseSpeech,
  parseWitchAction,
} from '../parser';
import * as prompts from '../prompts';

// ─── 夜晚阶段 ──────────────────────────────────────────────────

/**
 * 守卫守护。每晚先于狼人行动。
 * 不能连续两晚守同一人（player.guardLastProtected 记录）。
 */
export async function phaseGuardProtect(engine: GameEngine): Promise<void> {
  const log = engine.log;
  const guard = findAlivePlayer(engine, 'guard');
  if (!guard) return;

  // 守卫先于其他角色行动，需要先初始化当晚记录
  engine.ensureCurrentNight(log.day);

  // 守卫可以守任何人（包括自己），但不能守上一次守的
  const allIds = engine.players
    .filter((p) => p.alive && p.id !== guard.state.guardLastProtected)
    .map((p) => p.state.playerId);

  if (!allIds.length) return;

  const nightHistory = engine.nightHistorySummary();
  const { sys, user } = splitPrompt(
    prompts.guardProtectPrompt(
      guard.state,
      engine.allStates(),
      nightHistory,
      guard.state.personality,
    ),
  );

  const { text, usage } = await guard.chatWithUsage(sys, user);
  engine.recordLlmCall(guard, usage);

  let target = parseAction(text, 'protect', allIds);
  if (target === null && engine.dryRun) {
    target = allIds[Math.floor(Math.random() * allIds.length)];
  }

  if (target !== null) {
    engine.currentNight!.guardProtected = target;
    guard.state.guardLastProtected = target;
    const targetName = engine.getPlayerName(target);
    engine.emit('guard_protect', {
      playerId: guard.id,
      playerName: guard.name,
      targetId: target,
      targetName,
    }, true);

    if (engine.verbose) {
      console.info(`  🛡️  守卫${guard.name}守护: ${targetName}(${target})`);
    }
  }
}

export async function phaseWolfKill(engine: GameEngine): Promise<void> {
  const log = engine.log;
  const day = log.day;
  const night = engine.ensureCurrentNight(day);

  const wolves = engine.players.filter(
    (p) => p.alive && p.role === 'werewolf',
  );
  if (!wolves.length) return;

  const aliveIds = engine.players
    .filter((p) => p.alive && p.role !== 'werewolf')
    .map((p) => p.state.playerId);
  if (!aliveIds.length) return;

  const nightHistory = engine.nightHistorySummary();
  const votes = new Map<number, number>();

  for (const wolf of wolves) {
    const { sys, user } = splitPrompt(
      prompts.wolfKillPrompt(
        wolf.state,
        engine.allStates(),
        nightHistory,
        wolf.state.personality,
      ),
    );
    const { text, usage } = await wolf.chatWithUsage(sys, user);
    engine.recordLlmCall(wolf, usage);

    let target = parseAction(text, 'kill', aliveIds);
    if (target === null && engine.dryRun) {
      target = aliveIds[Math.floor(Math.random() * aliveIds.length)];
    }

    if (target !== null) {
      votes.set(wolf.id, target);
      engine.emit(
        'wolf_kill_decision',
        {
          playerId: wolf.id,
          playerName: wolf.name,
          targetId: target,
          targetName: engine.getPlayerName(target),
        },
        true,
      );
    }

    if (engine.verbose) {
      console.info(`  🐺 ${wolf.name} 选择杀: ${target}`);
    }
  }

  // 多数决（平局随机）
  if (votes.size > 0) {
    const tally = new Map<number, number>();
    for (const target of votes.values()) {
      tally.set(target, (tally.get(target) ?? 0) + 1);
    }
    const maxCount = Math.max(...tally.values());
    const candidates = [...tally.entries()]
      .filter(([, c]) => c === maxCount)
      .map(([t]) => t);
    night.wolfTarget = candidates[Math.floor(Math.random() * candidates.length)];
  } else {
    night.wolfTarget = aliveIds[Math.floor(Math.random() * aliveIds.length)];
  }

  engine.emit(
    'wolf_kill_result',
    {
      targetId: night.wolfTarget,
      targetName: engine.getPlayerName(night.wolfTarget!),
    },
    true,
  );

  if (engine.verbose) {
    const name = engine.getPlayerName(night.wolfTarget!);
    console.info(`  🌙 第${day}夜 狼人决定杀害: ${name}(${night.wolfTarget})`);
  }
}

export async function phaseSeerCheck(engine: GameEngine): Promise<void> {
  const night = engine.currentNight;
  if (!night) return;

  const seer = findAlivePlayer(engine, 'seer');
  if (!seer) return;

  const aliveIds = engine.players
    .filter((p) => p.alive && p.id !== seer.id)
    .map((p) => p.state.playerId);

  const { sys, user } = splitPrompt(
    prompts.seerCheckPrompt(
      seer.state,
      engine.allStates(),
      engine.seerResults,
      [],
      seer.state.personality,
    ),
  );
  const { text, usage } = await seer.chatWithUsage(sys, user);
  engine.recordLlmCall(seer, usage);

  let target = parseAction(text, 'check', aliveIds);
  if (target === null && engine.dryRun && aliveIds.length) {
    target = aliveIds[Math.floor(Math.random() * aliveIds.length)];
  }

  if (target !== null) {
    night.seerTarget = target;
    const targetPlayer = engine.getPlayerStateById(target);
    if (targetPlayer) {
      const isWolf = targetPlayer.role === 'werewolf';
      const result = isWolf ? 'wolf' : 'good';
      night.seerResult = result;
      engine.seerResults.push({
        id: target,
        name: targetPlayer.name,
        isWolf,
      });

      engine.emit(
        'seer_check',
        {
          seerId: seer.id,
          seerName: seer.name,
          targetId: target,
          targetName: targetPlayer.name,
          result,
        },
        true,
      );

      if (engine.verbose) {
        console.info(
          `  🔮 预言家${seer.name}查验: ${targetPlayer.name}(${target}) → ${isWolf ? '狼人' : '好人'}`,
        );
      }
    }
  }
}

export async function phaseWitchAction(engine: GameEngine): Promise<void> {
  const night = engine.currentNight;
  if (!night) return;

  const witch = findAlivePlayer(engine, 'witch');
  if (!witch) return;

  const hasAntidote = witch.state.witchHasAntidote;
  const hasPoison = witch.state.witchHasPoison;
  if (!hasAntidote && !hasPoison) return;

  const killedId = night.wolfTarget;
  const killedName = killedId !== null ? engine.getPlayerName(killedId) : '未知';

  const { sys, user } = splitPrompt(
    prompts.witchActionPrompt(
      witch.state,
      engine.allStates(),
      killedName,
      killedId,
      hasAntidote,
      hasPoison,
      [],
      witch.state.personality,
    ),
  );
  const { text, usage } = await witch.chatWithUsage(sys, user);
  engine.recordLlmCall(witch, usage);

  const action = parseWitchAction(text);

  let savedTargetId: number | null = null;
  let savedTargetName: string | null = null;
  let poisonTargetId: number | null = null;
  let poisonTargetName: string | null = null;

  // 救人
  if (action.save && hasAntidote && killedId !== null) {
    night.witchSaved = true;
    witch.state.witchHasAntidote = false;
    savedTargetId = killedId;
    savedTargetName = killedName;
    if (engine.verbose) {
      console.info(`  💊 女巫${witch.name}使用解药救了 ${killedName}`);
    }
  }

  // 毒人
  const aliveIds = engine.players
    .filter((p) => p.alive && p.id !== witch.id)
    .map((p) => p.state.playerId);
  const poisonRaw = action.poison;
  if (poisonRaw !== null && hasPoison && aliveIds.includes(poisonRaw)) {
    night.witchPoisoned = poisonRaw;
    witch.state.witchHasPoison = false;
    poisonTargetId = poisonRaw;
    poisonTargetName = engine.getPlayerName(poisonRaw);
    if (engine.verbose) {
      console.info(`  ☠️ 女巫${witch.name}使用毒药毒了 ${poisonTargetName}`);
    }
  }

  engine.emit(
    'witch_action',
    {
      witchId: witch.id,
      witchName: witch.name,
      saved: action.save && !!savedTargetId,
      savedTargetId,
      savedTargetName,
      poisonTargetId,
      poisonTargetName,
    },
    true,
  );
}

export function phaseNightResolve(engine: GameEngine): DeathRecord[] {
  const night = engine.currentNight;
  if (!night) return [];

  const deaths: DeathRecord[] = [];

  // 狼人杀人（除非被守卫挡 / 被女巫救）
  if (night.wolfTarget !== null && !night.witchSaved) {
    const protectedByGuard = night.guardProtected === night.wolfTarget;
    if (!protectedByGuard) {
      const target = engine.getPlayerStateById(night.wolfTarget);
      if (target && target.alive) {
        target.alive = false;
        deaths.push({
          playerId: night.wolfTarget,
          reason: DEATH_REASON.WOLF_KILL,
          day: night.day,
          details: '被狼人杀害',
        });
      }
    } else if (engine.verbose) {
      console.info(
        `  🛡️  守卫成功守护了 ${engine.getPlayerName(night.wolfTarget)}`,
      );
    }
  }

  // 女巫毒杀
  if (night.witchPoisoned !== null) {
    const target = engine.getPlayerStateById(night.witchPoisoned);
    if (target && target.alive) {
      target.alive = false;
      deaths.push({
        playerId: night.witchPoisoned,
        reason: DEATH_REASON.WITCH_POISON,
        day: night.day,
        details: '被女巫毒杀',
      });
    }
  }

  night.deaths = deaths;
  engine.log.nights.push(night);
  engine.log.deaths.push(...deaths);

  engine.emit(
    'night_deaths',
    {
      deaths: deaths.map((d) => ({
        playerId: d.playerId,
        name: engine.getPlayerName(d.playerId),
        reason: d.reason,
        details: d.details,
      })),
    },
    false,
  );

  return deaths;
}

// ─── 白天阶段 ──────────────────────────────────────────────────

export function phaseDayAnnounce(
  engine: GameEngine,
  deaths: DeathRecord[],
): string {
  let msg: string;
  if (deaths.length) {
    const names = deaths
      .map((d) => `${engine.getPlayerName(d.playerId)}(${d.details})`)
      .join('、');
    msg = `昨晚，${names}。`;
  } else {
    msg = '昨晚是平安夜，没有人死亡。';
  }

  engine.emit('day_announce', { deathsInfo: msg }, false);

  if (engine.verbose) {
    console.info(`  ☀️  ${msg}`);
  }
  return msg;
}

export async function phaseHunterShoot(
  engine: GameEngine,
  deaths: DeathRecord[],
): Promise<number | null> {
  for (const death of deaths) {
    const player = engine.getPlayerStateById(death.playerId);
    if (player && player.role === 'hunter' && player.hunterCanShoot) {
      const hunterPlayer = findPlayerById(engine, death.playerId);
      if (!hunterPlayer) continue;

      // 被女巫毒杀的猎人不能开枪
      if (death.reason === DEATH_REASON.WITCH_POISON) {
        if (engine.verbose) {
          console.info(`  🔫 猎人${player.name} 被毒杀，无法开枪`);
        }
        continue;
      }

      const aliveIds = engine.players.filter((p) => p.alive).map((p) => p.state.playerId);
      const reasonStr = death.details;

      const { sys, user } = splitPrompt(
        prompts.hunterShootPrompt(
          hunterPlayer.state,
          engine.allStates(),
          reasonStr,
          [],
          hunterPlayer.state.personality,
        ),
      );
      const { text, usage } = await hunterPlayer.chatWithUsage(sys, user);
      engine.recordLlmCall(hunterPlayer, usage);

      const target = parseAction(text, 'shoot', aliveIds);
      if (target !== null) {
        const targetPlayer = engine.getPlayerStateById(target);
        if (targetPlayer && targetPlayer.alive) {
          targetPlayer.alive = false;
          engine.log.deaths.push({
            playerId: target,
            reason: DEATH_REASON.HUNTER_SHOOT,
            day: engine.log.day,
            details: `被猎人${player.name}开枪射杀`,
          });
          engine.emit(
            'hunter_shoot',
            {
              hunterId: hunterPlayer.id,
              hunterName: hunterPlayer.name,
              targetId: target,
              targetName: targetPlayer.name,
            },
            false,
          );
          if (engine.verbose) {
            console.info(
              `  🔫 猎人${player.name}开枪射杀了 ${targetPlayer.name}(${target})`,
            );
          }
          return target;
        }
      }
    }
  }
  return null;
}

export async function phaseSpeech(
  engine: GameEngine,
  day: number,
  deathsInfo: string,
): Promise<Array<{ playerId: number; name: string; content: string }>> {
  const speeches: Array<{ playerId: number; name: string; content: string }> = [];
  const alivePlayers = engine.players.filter((p) => p.alive);

  for (const player of alivePlayers) {
    let personalInfo = '';
    if (player.role === 'seer') {
      personalInfo = prompts.seerPersonalInfo(engine.seerResults);
    } else if (player.role === 'werewolf') {
      const teammates = engine.players
        .filter(
          (p) =>
            p.role === 'werewolf' &&
            p.id !== player.id &&
            p.alive,
        )
        .map((p) => p.state);
      personalInfo = prompts.wolfPersonalInfo(teammates);
    }

    const { sys, user } = splitPrompt(
      prompts.speechPrompt(
        player.state,
        engine.allStates(),
        day,
        deathsInfo,
        speeches,
        personalInfo,
        player.state.personality,
      ),
    );
    const { text, usage } = await player.chatWithUsage(sys, user);
    engine.recordLlmCall(player, usage);

    const speechText = parseSpeech(text);
    speeches.push({
      playerId: player.id,
      name: player.name,
      content: speechText,
    });

    engine.emit(
      'speech',
      {
        playerId: player.id,
        playerName: player.name,
        content: speechText,
      },
      false,
    );

    if (engine.verbose) {
      console.info(
        `  🎤 [${player.id}] ${player.name}: ${speechText.slice(0, 80)}...`,
      );
    }
  }

  engine.log.speeches.push(speeches);
  return speeches;
}

export async function phaseVote(
  engine: GameEngine,
  day: number,
  speeches: Array<{ playerId: number; name: string; content: string }>,
): Promise<VoteResult> {
  const alivePlayers = engine.players.filter((p) => p.alive);
  const aliveIds = alivePlayers.map((p) => p.state.playerId);

  const votes: Record<number, number> = {};
  for (const player of alivePlayers) {
    let personalInfo = '';
    if (player.role === 'seer') {
      personalInfo = prompts.seerPersonalInfo(engine.seerResults);
    } else if (player.role === 'werewolf') {
      const teammates = engine.players
        .filter(
          (p) =>
            p.role === 'werewolf' &&
            p.id !== player.id &&
            p.alive,
        )
        .map((p) => p.state);
      personalInfo = prompts.wolfPersonalInfo(teammates);
    }

    const { sys, user } = splitPrompt(
      prompts.votePrompt(
        player.state,
        engine.allStates(),
        day,
        speeches,
        personalInfo,
        player.state.personality,
      ),
    );
    const { text, usage } = await player.chatWithUsage(sys, user);
    engine.recordLlmCall(player, usage);

    const validTargets = aliveIds.filter((pid) => pid !== player.id);
    let target = parseAction(text, 'vote', validTargets);
    if (target === null && engine.dryRun) {
      target = validTargets[Math.floor(Math.random() * validTargets.length)];
    }

    if (target !== null) {
      votes[player.id] = target;
      engine.emit(
        'vote_cast',
        {
          voterId: player.id,
          voterName: player.name,
          targetId: target,
          targetName: engine.getPlayerName(target),
        },
        false,
      );
    }

    if (engine.verbose) {
      const targetName = target !== null ? engine.getPlayerName(target) : '无效';
      console.info(`  🗳️  [${player.id}] ${player.name} → 投票给 ${targetName}`);
    }
  }

  // 计票
  const tally: Record<number, number> = {};
  for (const target of Object.values(votes)) {
    tally[target] = (tally[target] ?? 0) + 1;
  }

  let eliminated: number | null = null;
  let isTie = false;
  const tallyValues = Object.values(tally);
  if (tallyValues.length) {
    const maxCount = Math.max(...tallyValues);
    const candidates = Object.entries(tally)
      .filter(([, c]) => c === maxCount)
      .map(([t]) => parseInt(t, 10));
    if (candidates.length === 1) {
      eliminated = candidates[0];
    } else {
      isTie = true;
      eliminated = candidates[Math.floor(Math.random() * candidates.length)];
    }
  }

  // 执行放逐
  let eliminatedName: string | null = null;
  if (eliminated !== null) {
    const targetPlayer = engine.getPlayerStateById(eliminated);
    if (targetPlayer) {
      targetPlayer.alive = false;
      const count = tally[eliminated] ?? 0;
      engine.log.deaths.push({
        playerId: eliminated,
        reason: DEATH_REASON.VOTE_OUT,
        day,
        details: `被投票放逐(${count}票)`,
      });
      eliminatedName = targetPlayer.name;
      if (engine.verbose) {
        console.info(
          `  ⚖️  ${eliminatedName}(${eliminated}) 被投票放逐 (${count}票)`,
        );
      }
    }
  }

  const result: VoteResult = {
    day,
    votes,
    tally,
    eliminated,
    isTie,
  };
  engine.log.votes.push(result);

  engine.emit(
    'vote_result',
    {
      tally,
      votes,
      eliminated,
      eliminatedName,
      isTie,
    },
    false,
  );

  return result;
}

// ─── 辅助 ─────────────────────────────────────────────────────

function findAlivePlayer(engine: GameEngine, role: string): AIPlayer | null {
  return engine.players.find((p) => p.alive && p.role === role) ?? null;
}

function findPlayerById(engine: GameEngine, pid: number): AIPlayer | null {
  return engine.players.find((p) => p.id === pid) ?? null;
}

function splitPrompt(p: { system: string; user: string }): { sys: string; user: string } {
  return { sys: p.system, user: p.user };
}
